File: app/api/inventory.py
"""Inventory API: adjustments, damage/loss/expiry, movements, alerts, count sessions."""
from flask import Blueprint, g, jsonify, request

from ..database import get_db, query
from ..utils.audit import log_action
from ..utils.security import login_required, permission_required
from ..utils.stock import adjust_stock

bp = Blueprint("inventory", __name__, url_prefix="/api/inventory")

ADJUST_REASONS = {"adjustment", "damage", "loss", "expired", "transfer", "count"}


@bp.post("/adjust")
@permission_required("inventory")
def adjust():
    """Apply a manual stock change. ``mode='set'`` sets an absolute quantity,
    otherwise ``change`` is added/subtracted."""
    data = request.get_json(silent=True) or {}
    pid = data.get("product_id")
    reason = data.get("reason", "adjustment")
    if reason not in ADJUST_REASONS:
        return jsonify({"error": "invalid reason"}), 400
    db = get_db()
    prod = db.execute("SELECT quantity, name_en FROM products WHERE id = ?", (pid,)).fetchone()
    if not prod:
        return jsonify({"error": "product not found"}), 404

    if data.get("mode") == "set":
        target = float(data.get("quantity") or 0)
        change = target - (prod["quantity"] or 0)
    else:
        change = float(data.get("change") or 0)
    balance = adjust_stock(db, pid, change, reason, "manual", None, note=data.get("note", ""))
    db.commit()
    log_action("inventory_adjust", "product", pid,
               f"{reason} {change:+g} -> {balance} ({prod['name_en']})")
    return jsonify({"ok": True, "balance": balance})


@bp.get("/movements")
@login_required
def movements():
    args = request.args
    where, params = ["1=1"], []
    if args.get("product_id"):
        where.append("m.product_id = ?")
        params.append(args["product_id"])
    if args.get("reason"):
        where.append("m.reason = ?")
        params.append(args["reason"])
    if args.get("date_from"):
        where.append("date(m.created_at) >= ?")
        params.append(args["date_from"])
    if args.get("date_to"):
        where.append("date(m.created_at) <= ?")
        params.append(args["date_to"])
    rows = query(
        f"SELECT m.*, p.name_en AS product_name, u.username FROM inventory_movements m "
        f"LEFT JOIN products p ON p.id = m.product_id "
        f"LEFT JOIN users u ON u.id = m.user_id "
        f"WHERE {' AND '.join(where)} ORDER BY m.id DESC LIMIT 500",
        params,
    )
    return jsonify({"movements": [dict(r) for r in rows]})


@bp.get("/alerts")
@login_required
def alerts():
    low = query(
        "SELECT id, name_en, barcode, quantity, min_stock, unit FROM products "
        "WHERE is_archived=0 AND quantity <= min_stock AND min_stock > 0 "
        "ORDER BY quantity ASC LIMIT 200"
    )
    out_of_stock = query(
        "SELECT id, name_en, barcode, quantity FROM products "
        "WHERE is_archived=0 AND quantity <= 0 ORDER BY name_en LIMIT 200"
    )
    expiring = query(
        "SELECT id, name_en, barcode, quantity, expiry_date FROM products "
        "WHERE is_archived=0 AND expiry_date IS NOT NULL AND expiry_date != '' "
        "AND date(expiry_date) <= date('now','localtime','+30 day') "
        "ORDER BY expiry_date ASC LIMIT 200"
    )
    return jsonify({
        "low_stock": [dict(r) for r in low],
        "out_of_stock": [dict(r) for r in out_of_stock],
        "expiring": [dict(r) for r in expiring],
    })


@bp.get("/value")
@login_required
def inventory_value():
    row = query(
        "SELECT COALESCE(SUM(quantity*purchase_price),0) AS cost_value, "
        "COALESCE(SUM(quantity*selling_price),0) AS retail_value, "
        "COALESCE(SUM(quantity),0) AS total_units, COUNT(*) AS products "
        "FROM products WHERE is_archived=0", one=True,
    )
    return jsonify(dict(row))


@bp.post("/count-session")
@permission_required("inventory")
def count_session():
    """Reconcile a physical count. ``counts`` = [{product_id, counted}]."""
    data = request.get_json(silent=True) or {}
    counts = data.get("counts") or []
    db = get_db()
    adjusted = 0
    for c in counts:
        pid = c.get("product_id")
        counted = float(c.get("counted") or 0)
        prod = db.execute("SELECT quantity FROM products WHERE id=?", (pid,)).fetchone()
        if not prod:
            continue
        diff = counted - (prod["quantity"] or 0)
        if diff != 0:
            adjust_stock(db, pid, diff, "count", "count_session", None,
                         note=f"Physical count reconciliation ({diff:+g})")
            adjusted += 1
    db.commit()
    log_action("inventory_count", detail=f"{adjusted} products reconciled")
    return jsonify({"ok": True, "adjusted": adjusted})

File: app/static/js/views/inventory.js
/* Inventory view — value, alerts, adjustments, movements, count sessions. */
KAKA_VIEWS.inventory = {
  title: () => t('nav.inventory'),
  async render(root) {
    UI.clear(root);
    const tabs = ['alerts', 'adjust', 'movements', 'count'];
    const labels = { alerts: 'Alerts', adjust: 'Adjust Stock', movements: 'Movements', count: 'Count Session' };
    let active = 'alerts';
    const tabBar = UI.el('div', { class: 'tabs' }, tabs.map(tb =>
      UI.el('div', { class: 'tab' + (tb === active ? ' active' : ''), text: labels[tb],
        onClick: () => { active = tb; document.querySelectorAll('.tab').forEach(x => x.classList.remove('active')); event.target.classList.add('active'); draw(); } })));
    const body = UI.el('div', {});
    root.appendChild(tabBar); root.appendChild(body);

    const val = await API.get('/api/inventory/value');
    root.insertBefore(UI.el('div', { class: 'grid stats-grid mb' }, [
      stat('Inventory (cost)', money(val.cost_value), val.products + ' products', 'blue'),
      stat('Inventory (retail)', money(val.retail_value), num(val.total_units) + ' units', 'green'),
      stat('Potential profit', money(val.retail_value - val.cost_value), '', 'orange'),
    ]), tabBar);

    draw();
    async function draw() {
      UI.clear(body);
      if (active === 'alerts') return drawAlerts();
      if (active === 'adjust') return drawAdjust();
      if (active === 'movements') return drawMovements();
      if (active === 'count') return drawCount();
    }

    async function drawAlerts() {
      const d = await API.get('/api/inventory/alerts');
      body.appendChild(UI.el('div', { class: 'card mb' }, [
        UI.el('h3', { class: 'mb', text: '⚠ Low Stock (' + d.low_stock.length + ')' }),
        UI.table([
          { key: 'name_en', label: t('common.name') },
          { key: 'barcode', label: t('prod.barcode') },
          { key: 'quantity', label: t('prod.stock'), num: true, render: r => num(r.quantity) },
          { key: 'min_stock', label: 'Min', num: true, render: r => num(r.min_stock) },
        ], d.low_stock, { empty: 'All good' })]));
      body.appendChild(UI.el('div', { class: 'card mb' }, [
        UI.el('h3', { class: 'mb', text: '⛔ Out of Stock (' + d.out_of_stock.length + ')' }),
        UI.table([{ key: 'name_en', label: t('common.name') }, { key: 'barcode', label: t('prod.barcode') }], d.out_of_stock, { empty: 'None' })]));
      body.appendChild(UI.el('div', { class: 'card' }, [
        UI.el('h3', { class: 'mb', text: '📅 Expiring within 30 days (' + d.expiring.length + ')' }),
        UI.table([{ key: 'name_en', label: t('common.name') }, { key: 'expiry_date', label: 'Expiry' },
          { key: 'quantity', label: t('prod.stock'), num: true, render: r => num(r.quantity) }], d.expiring, { empty: 'None' })]));
    }

    async function drawAdjust() {
      const search = UI.el('input', { class: 'grow', placeholder: 'Search product to adjust...' });
      const results = UI.el('div', {});
      body.appendChild(UI.el('div', { class: 'card' }, [
        UI.el('div', { class: 'toolbar' }, [search]), results]));
      let timer;
      search.addEventListener('input', () => { clearTimeout(timer); timer = setTimeout(async () => {
        if (!search.value.trim()) { UI.clear(results); return; }
        const d = await API.get('/api/products/search?q=' + encodeURIComponent(search.value.trim()));
        UI.clear(results);
        results.appendChild(UI.table([
          { key: 'name_en', label: t('common.name') },
          { key: 'quantity', label: t('prod.stock'), num: true, render: r => num(r.quantity) },
          { key: 'act', label: '', num: true, render: r => UI.el('button', { class: 'btn btn-sm btn-primary', text: 'Adjust', onClick: () => adjustModal(r) }) },
        ], d.products, { empty: 'No match' }));
      }, 200); });
    }
    function adjustModal(p) {
      const reason = UI.el('select', {}, ['adjustment', 'damage', 'loss', 'expired', 'transfer'].map(r => UI.el('option', { value: r }, r)));
      const mode = UI.el('select', {}, [UI.el('option', { value: 'change' }, 'Add / Remove (±)'), UI.el('option', { value: 'set' }, 'Set exact quantity')]);
      const qty = UI.el('input', { type: 'number', step: 'any', value: 0 });
      const note = UI.el('input', { placeholder: 'Note (optional)' });
      const body2 = UI.el('div', {}, [
        UI.el('p', { class: 'mb', html: `<strong>${UI.esc(p.name_en)}</strong> — current: ${num(p.quantity)}` }),
        UI.el('div', { class: 'field' }, [UI.el('label', { text: 'Reason' }), reason]),
        UI.el('div', { class: 'field mt' }, [UI.el('label', { text: 'Mode' }), mode]),
        UI.el('div', { class: 'field mt' }, [UI.el('label', { text: 'Quantity' }), qty]),
        UI.el('div', { class: 'field mt' }, [UI.el('label', { text: 'Note' }), note]),
      ]);
      UI.modal('Adjust Stock', body2, [
        { label: t('common.cancel'), class: 'btn-ghost' },
        { label: t('common.save'), class: 'btn-primary', onClick: async () => {
          await API.post('/api/inventory/adjust', { product_id: p.id, reason: reason.value,
            mode: mode.value, change: parseFloat(qty.value) || 0, quantity: parseFloat(qty.value) || 0, note: note.value });
          UI.toast('Adjusted', 'success'); draw();
        } }]);
    }

    async function drawMovements() {
      const d = await API.get('/api/inventory/movements');
      body.appendChild(UI.el('div', { class: 'card' }, [
        UI.el('h3', { class: 'mb', text: 'Recent Movements' }),
        UI.table([
          { key: 'created_at', label: t('common.date') },
          { key: 'product_name', label: t('common.name') },
          { key: 'reason', label: 'Reason', render: r => UI.badge(r.reason, 'blue') },
          { key: 'change_qty', label: 'Change', num: true, render: r => (r.change_qty > 0 ? '+' : '') + num(r.change_qty, 2) },
          { key: 'balance', label: 'Balance', num: true, render: r => num(r.balance, 2) },
          { key: 'username', label: 'By' },
        ], d.movements, { empty: 'No movements' })]));
    }

    async function drawCount() {
      const d = await API.get('/api/products?per_page=200');
      const counts = {};
      const rows = d.products.map(p => ({ p, input: (() => {
        const i = UI.el('input', { type: 'number', step: 'any', value: p.quantity, style: 'width:90px' });
        i.addEventListener('change', () => { counts[p.id] = parseFloat(i.value) || 0; });
        counts[p.id] = p.quantity; return i; })() }));
      body.appendChild(UI.el('div', { class: 'card' }, [
        UI.el('div', { class: 'section-head' }, [UI.el('h3', { text: 'Physical Count' }),
          UI.el('button', { class: 'btn btn-primary', text: 'Reconcile', onClick: async () => {
            const payload = Object.entries(counts).map(([id, c]) => ({ product_id: +id, counted: c }));
            const r = await API.post('/api/inventory/count-session', { counts: payload });
            UI.toast(`Reconciled ${r.adjusted} products`, 'success'); draw();
          } })]),
        UI.table([
          { key: 'name', label: t('common.name'), render: r => r.p.name_en },
          { key: 'system', label: 'System Qty', num: true, render: r => num(r.p.quantity) },
          { key: 'counted', label: 'Counted', num: true, render: r => r.input },
        ], rows)]));
    }
  },
};
